x-pack/plugins/apm/scripts/create-functional-tests-archive/index.ts

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { argv } from 'yargs';
+import { execSync } from 'child_process';
+import moment from 'moment';
+import path from 'path';
+import fs from 'fs';
+import { stampLogger } from '../shared/stamp-logger';
+
+async function run() {
+  stampLogger();
+
+  const archiveName = 'apm_8.0.0';
+
+  // include important APM data and ML data
+  const indices =
+    'apm-*-transaction,apm-*-span,apm-*-error,apm-*-metric,.ml-anomalies*,.ml-config';
+
+  const esUrl = argv['es-url'] as string | undefined;
+
+  if (!esUrl) {
+    throw new Error('--es-url is not set');
+  }
+  const kibanaUrl = argv['kibana-url'] as string | undefined;
+
+  if (!kibanaUrl) {
+    throw new Error('--kibana-url is not set');
+  }
+  const gte = moment().subtract(1, 'hour').toISOString();
+  const lt = moment(gte).add(30, 'minutes').toISOString();
+
+  // eslint-disable-next-line no-console
+  console.log(`Archiving from ${gte} to ${lt}...`);
+
+  // APM data uses '@timestamp' (ECS), ML data uses 'timestamp'
+
+  const rangeQueries = [
+    {
+      range: {
+        '@timestamp': {
+          gte,
+          lt,
+        },
+      },
+    },
+    {
+      range: {
+        timestamp: {
+          gte,
+          lt,
+        },
+      },
+    },
+  ];
+
+  // some of the data is timeless/content
+  const query = {
+    bool: {
+      should: [
+        ...rangeQueries,
+        {
+          bool: {
+            must_not: [
+              {
+                exists: {
+                  field: '@timestamp',
+                },
+              },
+              {
+                exists: {
+                  field: 'timestamp',
+                },
+              },
+            ],
+          },
+        },
+      ],
+      minimum_should_match: 1,
+    },
+  };
```

Comment:
This is probably more a comment for https://github.com/elastic/kibana/pull/76522 but only noticing this now: Is it not possible to specify a `query` per index?
I see that esarchiver takes `indices` as a separate argument eg `node scripts/es_archiver save ${archiveName} ${indices}`. Would it make sense to treat this as a shorthand if no `--query` is given but ignore it if a query is given, and then require the query to contain the index?

So you query above would be:
```js
const rangeQueries = [
  {
    index: 'apm-*-transaction,apm-*-span,apm-*-error,apm-*-metric',
    body: { range: { "@timestamp": { gte, lt } } }}
  },
  {
    index: '.ml-anomalies*,.ml-config',
    body: { range: { timestamp: { gte, lt } } }
  }
];
```

That would make this a whole lot simpler.

Reply:
I think that's incompatible with how esarchiver works today: it will scroll through the hits for one search. It doesn't create a different search for every index.